README.md
[![NPM version](https://img.shields.io/npm/v/express-source-fragment.svg)](https://www.npmjs.com/package/express-source-fragment)
[![Build Status](https://travis-ci.org/lahmatiy/express-source-fragment.svg?branch=master)](https://travis-ci.org/lahmatiy/express-source-fragment)
[![Coverage Status](https://coveralls.io/repos/github/lahmatiy/express-source-fragment/badge.svg?branch=master)](https://coveralls.io/github/lahmatiy/express-source-fragment?branch=master)
[![Dependency Status](https://img.shields.io/david/lahmatiy/express-source-fragment.svg)](https://david-dm.org/lahmatiy/express-source-fragment)

Express middleware to get a file content fragment by request to defined route. Based on [source-fragment](https://github.com/lahmatiy/source-fragment).

## Install

```
npm install express-source-fragment
```

## Usage

```js
var express = require('express');
var sourceFragment = require('express-source-fragment');

var app = express();

// There are few ways to setup:
// - to trigger middleware on *GET* request to `/source-fragment`
app.get('/source-fragment', sourceFragment());

// - to trigger middleware on *any* request method to `/source-fragment`
app.use('/source-fragment', sourceFragment());

// - to trigger middleware on *any* request method to *any* path
//   (not recommended unless server's single purpose is a serving of source fragments)
app.use(sourceFragment());
```

After that you can use `GET` requests like `/source-fragment?file=foo/bar.js:2:5:4:12` to get a source fragment of `foo/bar.js` that starts on line 2 column 5 and ends on line 4 column 12.

### Using with webpack-dev-server

Although `webpack-dev-server` uses `express` to create a server, you have the same options to apply the middleware to it. The only difference is that you should define it in `setup` method (see [issue](https://github.com/webpack/webpack-dev-server/issues/285) for details).

```js
var server = new WebpackDevServer(webpack(config), {
  // ...
  setup: function(app) {
    app.use('/source-fragment', sourceFragment());
  }
});
```

## API

```
sourceFragment([options]);
```

Options:

- `cwd`

  Type: `String`  
  Default: `''` (uses `process.cwd()`)

  Working directory used when resolving paths.

## Related projects

- [source-fragment](https://github.com/lahmatiy/source-fragment) – package that do the main task of `express-source-fragment`, i.e. gets a source fragment.
- [express-open-in-editor](https://github.com/lahmatiy/express-open-in-editor) – Express middleware to open a file in an editor by request, [open-in-editor](https://github.com/lahmatiy/open-in-editor).
- [babel-plugin-source-wrapper](https://github.com/restrry/babel-plugin-source-wrapper) – `Babel` plugin that instruments source code to associate objects with location they defined in code base.
- [Component Inspector](https://github.com/lahmatiy/component-inspector) – developer tool to inspect components that can open component creation source location in editor. Has integrations for `React`, `Backbone` and can be adopter for other frameworks.

## License

MIT


HISTORY.md
## 1.1.0 (October 26, 2017)

- Bump `source-fragment` version to `^1.1.0`
- Changed output of source fragments to shift code to the left by removing spaces as much as possible (using `collapseOffset` option)

## 1.0.0 (October 1, 2017)

- Initial release


test/index.js
var assert = require('assert');
var express = require('express');
var http = require('http');
var path = require('path');
var sourceFragment = require('../index.js');
var PORT = process.env.PORT || 8999;

function request(options, expectedStatusCode, expectedResponse, done) {
    return function(done) {
        http.request(Object.assign({
            hostname: 'localhost',
            port: PORT,
            method: 'GET'
        }, options), function(res) {
            var actualResponse = '';

            assert.equal(res.statusCode, expectedStatusCode);

            if (expectedStatusCode === 200) {
                assert.equal(res.headers['content-type'], 'text/html; charset=utf-8');
            }

            res.setEncoding('utf8');
            res.on('data', function(chunk) {
                actualResponse += chunk;
            });
            res.on('end', function() {
                if (typeof expectedResponse === 'string') {
                    assert.equal(actualResponse, expectedResponse);
                } else {
                    assert(expectedResponse.test(actualResponse), 'should match to ' + expectedResponse);
                }
                done();
            });
        }).end();
    };
}

describe('app.use(middleware)', function() {
    var server;
    var app;
    var tests = [
        [{ path: '/any?loc=test/fixture/script.js:2:12:5:6' }, 200, /^<div [^>]+>(\s|.)+?<\/div>$/],
        [{ path: '/any?loc=test/fixture/bad-path.js:2:12:5:6' }, 404, '[express-source-fragment] File not found: ' + path.join(process.cwd(), '/test/fixture/bad-path.js:2:12:5:6') + ' (base: ' + process.cwd() + ', requested: test/fixture/bad-path.js:2:12:5:6)'],
        [{ path: '/any?loc=test/fixture/script.js:2:12:5:6', method: 'POST' }, 200, /^<div [^>]+>(\s|.)+?<\/div>$/],
        [{ path: '/any?locx=test/fixture/script.js:2:12:5:6' }, 400, '[express-source-fragment] Parameter missed: loc']
    ];

    beforeEach(function(done) {
        app = express();
        app.use(sourceFragment());
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    tests.forEach(function(test) {
        var options = test[0];
        var statusCode = test[1];
        var response = test[2];

        it([
            options.method || 'GET',
            options.path
        ].join(' '), request(options, statusCode, response));
    });
});

describe('app.use("path", middleware)', function() {
    var server;
    var app;
    var tests = [
        [{ path: '/test?loc=test/fixture/script.js:2:12:5:6' }, 200, /^<div [^>]+>(\s|.)+?<\/div>$/],
        [{ path: '/test?loc=test/fixture/bad-path.js:2:12:5:6' }, 404, '[express-source-fragment] File not found: ' + path.join(process.cwd(), '/test/fixture/bad-path.js:2:12:5:6') + ' (base: ' + process.cwd() + ', requested: test/fixture/bad-path.js:2:12:5:6)'],
        [{ path: '/test?loc=test/fixture/script.js:2:12:5:6', method: 'POST' }, 200, /^<div [^>]+>(\s|.)+?<\/div>$/],
        [{ path: '/test?locx=test/fixture/script.js:2:12:5:6' }, 400, '[express-source-fragment] Parameter missed: loc'],
        [{ path: '/?loc=test/fixture/script.js:2:12:5:6' }, 404, /Cannot GET \//],
        [{ path: '/xx?loc=test/fixture/script.js:2:12:5:6' }, 404, /Cannot GET \/xx/]
    ];

    beforeEach(function(done) {
        app = express();
        app.use('/test', sourceFragment());
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    tests.forEach(function(test) {
        var options = test[0];
        var statusCode = test[1];
        var response = test[2];

        it([
            options.method || 'GET',
            options.path
        ].join(' '), request(options, statusCode, response));
    });
});

describe('app.get("path", middleware)', function() {
    var server;
    var app;
    var tests = [
        [{ path: '/test?loc=test/fixture/script.js:2:12:5:6' }, 200, /^<div [^>]+>(\s|.)+?<\/div>$/],
        [{ path: '/test?loc=test/fixture/bad-path.js:2:12:5:6' }, 404, '[express-source-fragment] File not found: ' + path.join(process.cwd(), '/test/fixture/bad-path.js:2:12:5:6') + ' (base: ' + process.cwd() + ', requested: test/fixture/bad-path.js:2:12:5:6)'],
        [{ path: '/test?loc=test/fixture/script.js:2:12:5:6', method: 'POST' }, 404, /Cannot POST \/test/],
        [{ path: '/test?locx=test/fixture/script.js:2:12:5:6' }, 400, '[express-source-fragment] Parameter missed: loc'],
        [{ path: '/?loc=test/fixture/script.js:2:12:5:6' }, 404, /Cannot GET \//],
        [{ path: '/xx?loc=test/fixture/script.js:2:12:5:6' }, 404, /Cannot GET \/xx/]
    ];

    beforeEach(function(done) {
        app = express();
        app.get('/test', sourceFragment());
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    tests.forEach(function(test) {
        var options = test[0];
        var statusCode = test[1];
        var response = test[2];

        it([
            options.method || 'GET',
            options.path
        ].join(' '), request(options, statusCode, response));
    });
});

describe('options', function() {
    var server;
    var app;
    var tests = [
        [{ path: '/defaultCwd?loc=test/fixture/script.js:2:12:5:6' }, 200, /^<div [^>]+>(\s|.)+?<\/div>$/],
        [{ path: '/defaultCwd?loc=script.js:2:12:5:6' }, 404, '[express-source-fragment] File not found: ' + path.join(process.cwd(), '/script.js:2:12:5:6') + ' (base: ' + process.cwd() + ', requested: script.js:2:12:5:6)'],
        [{ path: '/fixture?loc=script.js:2:12:5:6' }, 200, /^<div [^>]+>(\s|.)+?<\/div>$/],
        [{ path: '/fixture?loc=test/fixture/script.js:2:12:5:6' }, 404, '[express-source-fragment] File not found: ' + path.join(process.cwd(), 'test/fixture/', '/test/fixture/script.js:2:12:5:6') + ' (base: ' + path.join(process.cwd(), 'test/fixture') + ', requested: test/fixture/script.js:2:12:5:6)']
    ];

    beforeEach(function(done) {
        app = express();
        app.get('/defaultCwd', sourceFragment());
        app.get('/fixture', sourceFragment({
            cwd: 'test/fixture'
        }));
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    tests.forEach(function(test) {
        var options = test[0];
        var statusCode = test[1];
        var response = test[2];

        it([
            options.method || 'GET',
            options.path
        ].join(' '), request(options, statusCode, response));
    });
});


index.js
var url = require('url');
var fs = require('fs');
var path = require('path');
var sf = require('source-fragment');
var MESSAGE_PREFIX = '[' + require('./package.json').name + '] ';

function fail(res, code, message) {
    res.statusCode = code;
    res.end(MESSAGE_PREFIX + message);
}

module.exports = function(options) {
    var cwd = path.resolve((options && options.cwd) || process.cwd());
    return function sourceFragment(req, res, next) {
        var queryLoc = url.parse(req.url, true).query.loc;
        var loc;

        if (!queryLoc) {
            return fail(res, 400, 'Parameter missed: loc');
        }

        // resolve to absolute path
        loc = path.join(cwd, path.resolve('/', queryLoc));

        if (!fs.existsSync(loc.replace(/(:\d+)+$/, ''))) {
            return fail(res, 404,
                'File not found: ' + loc + ' (base: ' + cwd + ', requested: ' + queryLoc + ')'
            );
        }

        try {
            res.statusCode = 200;
            res.set('Content-Type', 'text/html; charset=utf-8');
            res.end(sf(loc, {
                format: 'html',
                collapseOffset: true
            }));
        } catch (e) {
            fail(res, 500, 'ERROR: ' + e);
        }
    };
};
